internal/jsonwall/jsonwall.go

```diff
@@ -0,0 +1,307 @@
+//
+// The entries in this database might be manipulated with a type such as:
+//
+//     type AppEntry struct {
+//             Kind string    `json:"kind"`
+//             Name string    `json:"name,omitempty"`
+//             Version string `json:"version,omitempty"`
+//     }
+//
+// Such data types have two important characteristics: fields must be defined in
+// the order that will be used when searching, and every optional field must be
+// tagged with `omitempty`.
+//
+// With that in place, the database may be accessed as:
+//
+//     app := AppEntry{Kind: "app", Name: "chisel"}
+//     if db.Get(&app) == nil {
+//             fmt.Println(app.Name, "version:", app.Version)
+//     }
+//
+// Iteration works similarly:
+//
+//     app := AppEntry{Kind: "app"}
+//     if iter, err := db.Iter(&app); err == nil {
+//             for iter.Next() {
+//                     if iter.Get(&app) == nil {
+//                             fmt.Println(app.Name, "version:", app.Version)
+//                     }
+//             }
+//     }
+//
+package jsonwall
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sort"
+	"strings"
+)
+
+// DBWriter holds in memory the state of a database while it's being prepared
+// for serialization and implements the WriterTo interface for assembling it.
+type DBWriter struct {
+	options *DBWriterOptions
+	entries [][]byte
+}
+
+type DBWriterOptions struct {
+	// Schema is included in the database header to help the decoding
+	// process. The value is made available when reading, and is not
+	// internally interpreted.
+	Schema string
+}
+
+// NewDBWriter returns a database writer that can assemble new databases.
+func NewDBWriter(options *DBWriterOptions) *DBWriter {
+	if options == nil {
+		options = &DBWriterOptions{}
+	}
+	return &DBWriter{options: options}
+}
+
+// Add encodes the provided value as a JSON object and includes the resulting
+// data into the database being created.
+func (dbw *DBWriter) Add(value any) error {
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	if len(data) == 0 || data[0] != '{' {
+		return fmt.Errorf("invalid database value: %#v", value)
+	}
+	dbw.entries = append(dbw.entries, data)
+	return nil
+}
+
+type sortableEntries [][]byte
+
+func (e sortableEntries) Less(i, j int) bool { return bytes.Compare(e[i], e[j]) < 0 }
+func (e sortableEntries) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
+func (e sortableEntries) Len() int           { return len(e) }
+
+type jsonwallHeader struct {
+	Version string `json:"jsonwall"`
+	Schema  string `json:"schema,omitempty"`
+	Count   int    `json:"count"`
+}
+
+const jsonwallVersion = "1.0"
+
+func (dbw *DBWriter) writeHeader(w io.Writer, count int) (int, error) {
+	data, err := json.Marshal(&jsonwallHeader{
+		Version: jsonwallVersion,
+		Schema:  dbw.options.Schema,
+		Count:   count,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("internal error: cannot marshal database header")
```

Comment:
Some of the other occurrences are intentionally hiding internal debug data from the user terminal when that's unnecessary. For the case here, it seems fine either way since this would be a major mistake somewhere. I'll add the extra debug data, thanks.